### ISSUE TITLE
Good Command system design

### ISSUE THREAD

**Issue by Snowiiii** [Snowiiii/Pumpkin#147]
- [x] I've searched existing issues and couldn't find a duplicate.

I'm thinking of improving the current design of our Command System which looks like this
![image](https://github.com/user-attachments/assets/260d2415-cf17-4232-a3fe-5e11988379a7)

One Problem for example is to no able to use Async Closures (yet).

Please let me know how do you think a Command System should look like?

**Comment by van-sprundel** [comment]
Would attribute macro's be something you'd be okay with? that's kind of the thing I would go for.

Something like
```rust
#[command(
    names = ["kick"],
    description = "Kicks the player from the server."
)]
struct KickCommand {
    target: PlayerName,
    reason: String,
}

impl CommandHandler for KickCommand {
    pub fn init_command_tree<'a>(...) -> CommandTree<'a> {
        ...
    }
}
```

**Comment by AnonymousBit0111** [comment]
What exactly is the use of the Command tree struct? I think we could just have the command handler execute the command and be done with it no?

**Comment by Snowiiii** [comment]
> What exactly is the use of the Command tree struct? I think we could just have the command handler execute the command and be done with it no?

Its pretty useful for tab completion and also a quick and easy way to check permission etc. Look at Mojang's [Brigadier](https://github.com/Mojang/brigadier) API

**Comment by AnonymousBit0111** [comment]
Yeah I see it now, I think if we use macros we'll need to have a way to handle arguments that require more arguments aswell, it needs to be evaluated in a chain before we execute anything

**Comment by suprohub** [comment]
Macro rules good too
like
```rust
command!{
names = ["kick"]
desc = "Kicks player"
{when goes comand tree
}
}
```
